doc/examples/server_selection.rst

```diff
@@ -0,0 +1,82 @@
+Server Selector Example
+=======================
+
+Users can exert fine-grained control over the `server selection algorithm`_
+by setting the `serverSelector` option on the :class:`~pymongo.MongoClient`
+to an appropriate callable. This example shows how to use this functionality
+to prefer servers running on ``localhost``.
+
+
+.. note::
+
+   Directing queries to servers running on ``localhost`` can be desirable
+   when using a sharded cluster with multiple ``mongos``, as locally run
+   queries are likely to see lower latency and higher throughput.
+
+
+.. testsetup::
+
+   from pymongo import MongoClient
+
+
+.. _server selection algorithm: https://docs.mongodb.com/manual/core/read-preference-mechanics/
+
+
+Example: Selecting Servers Running on ``localhost``
+---------------------------------------------------
+
+To start, we need to write the server selector function that will be used.
+The server selector function should accept a list of
+:class:`~pymongo.server_description.ServerDescription` objects and return a
+list of server descriptions that are suitable for the read or write operation.
+
+In addition to comparing the hostname with ``localhost``, our server selector
+function accounts for the following edge cases:
+
+* The list of server descriptions passed to the server selector is empty. This
+  can happen if the server selection process starts before the client has
+  received successful heartbeats from any of the ``mongod`` or ``mongos``.
```

Comment:
@jyemin is this how the Java selector works? Should we bother calling the custom selector with an empty list? I'd like us to work the same way across drivers.

Reply:
Discussed with Jeff offline, the Java driver does not call the selector if the list of servers is empty. Can you make python do the same?